sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp

```diff
@@ -1937,24 +1939,23 @@ pi_result piTearDown(void *) {
     }
     int Status = cm_support::CM_FAILURE;
 
-    if (Mem->getMemType() == PI_MEM_TYPE_BUFFER) {
+    if (Mem->SurfacePtr.tag == cm_surface_ptr_t::type_user_provided_buffer) {
       _pi_buffer *PiBuf = static_cast<_pi_buffer *>(Mem);
-      if (PiBuf->BufferPtr.tag == cm_buffer_ptr_slot::type_user_provided) {
-        Status = Mem->Context->Device->CmDevicePtr->DestroyBufferUP(
-            PiBuf->BufferPtr.UPBufPtr);
-      } else {
-        Status = Mem->Context->Device->CmDevicePtr->DestroySurface(
-            PiBuf->BufferPtr.RegularBufPtr);
-      }
-    } else if (Mem->getMemType() == PI_MEM_TYPE_IMAGE2D) {
+      Status = Mem->Context->Device->CmDevicePtr->DestroyBufferUP(
+          PiBuf->SurfacePtr.UPBufPtr);
+    } else if (Mem->SurfacePtr.tag == cm_surface_ptr_t::type_regular_buffer) {
+      _pi_buffer *PiBuf = static_cast<_pi_buffer *>(Mem);
+      Status = Mem->Context->Device->CmDevicePtr->DestroySurface(
```

Comment:
Could this functionality be encapsulated in destrutor of _pi_mem ?

Reply:
Done.